Plane: Introduced separate GCS failsafe to break it out from FS_LONG

mday299 opened the pull request:
Broke the GCS failsafe out into its own failure mode.  This keeps it separate from FS_LONG.  Future pull requests might consider not glomming all failure modes together into FS_LONG and FS_SHORT.  Rather keep throttle, battery, GCS, GPS, companion computer, engine RPMs, etc. all in their own failure modes.

WickedShell commented on ArduPlane/Parameters.h:
```diff
@@ -266,8 +266,8 @@ class Parameters {
         k_param_throttle_suppress_manual,
         k_param_throttle_passthru_stabilize,
         k_param_rc_12_old,
-        k_param_fs_batt_voltage, // unused - moved to AP_BattMonitor
```
You can't grab the old indexes as that breaks param conversion for people on older boards.

WickedShell commented on ArduPlane/Parameters.cpp:
```diff
@@ -1301,10 +1318,6 @@ const AP_Param::ConversionInfo conversion_table[] = {
     { Parameters::k_param_land_flap_percent,  0,      AP_PARAM_INT8,  "LAND_FLAP_PERCENT" },
 
-    // battery failsafes
-    { Parameters::k_param_fs_batt_voltage,    0,      AP_PARAM_FLOAT, "BATT_LOW_VOLT" },
```
We still need the conversion code to be retained.

WickedShell commented on ArduPlane/Parameters.cpp:
```diff
@@ -522,13 +522,30 @@ const AP_Param::Info Plane::var_info[] = {
+    // @Description: The action to take on a GCS failsafe event. If the aircraft was in a stabilization or manual mode when failsafe started and a GCS failsafe occurs then it will change to RTL mode if FS_GCS_ACTION is 0 or 1, and will change to FBWA if FS_GCS_ACTION is set to 2. If the aircraft was in an auto mode (such as AUTO or GUIDED) when the failsafe started then it will continue in the au ...
+    // @Values: 0:Continue,1:ReturnToLaunch,2:Glide,3:Deploy Parachute
+    // @User: Standard
+    GSCALAR(fs_action_gcs, "FS_GCS_ACTION", FS_ACTION_GCS_CONTINUE),
```
Would you be ameniable to using the FailsafeAction enum here and adding any other cases we don't currently have to the enum/handler for it? I'd like to see us use a more common and flexible set of actions for failsafes moving forwards.

mday299 replied:
Not sure I understand you. Are you talking about the FaiisafeAction enum in @peterbarker 's code?  If so, didn't you just say that's a third rail?  Of course, you could be speaking about another FailsafeAction enum.

WickedShell replied:
@mday299 Sorry I should have been a lot clearer. I meant specifically the actions in this enum: https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/Plane.h#L1034 currently they are used in the battery failsafe, but all of the listed actions I believe are quite valid for GCS failsafes, and any useful new ones would be probably be good for battery failsafe actions.

The reason this part isn't as bad to touch is because the list is unique per vehicle type at the moment, but external libraries can still interact with it.

WickedShell commented on ArduPlane/Parameters.cpp:
```diff
@@ -522,13 +522,30 @@ const AP_Param::Info Plane::var_info[] = {
+    // @DisplayName: GCS failsafe timeout
+    // @Description: The time in seconds that a GCS failsafe condition has to persist. This defaults to 0 seconds which means "disabled".  
+    // @Units: s
+    // @Range: 3 2147483647
```
We can lower the maximum range limit. Users can always set the value to whatever they want, but the range is used to generate slider objects normally in UI code, and this is a crazy scale for the slider.

WickedShell commented on ArduPlane/Plane.h:
```diff
@@ -909,8 +911,9 @@ class Plane : public AP_Vehicle {
     bool set_mode_by_number(const Mode::Number new_mode_number, const mode_reason_t reason);
     Mode *mode_from_mode_num(const enum Mode::Number num);
-    void check_long_failsafe();
+    void check_long_failsafe();  
```
Bad whitespace change?

magicrub commented:
The concept is good. There's been talk about abstracting all the fail-safes into it's own library for all vehicles. Perhaps @peterbarker already has a branch for it?

WickedShell commented on ArduPlane/events.cpp:
```diff
@@ -152,6 +152,79 @@ void Plane::failsafe_long_off_event(mode_reason_t reason)
+void Plane::failsafe_gcs_on_event(enum failsafe_state fstype, mode_reason_t reason)
+{
+    // This is how to handle a gcs loss of control signal failsafe.
+    gcs().send_text(MAV_SEVERITY_WARNING, "Failsafe. GCS event on: type=%u/reason=%u", fstype, reason);
```
No need to put the type in the print here, we already knew it was a GCS event.

WickedShell commented:
> There's been talk about abstracting all the fail-safes into it's own library for all vehicles. Perhaps @peterbarker already has a branch for it?

That's a third rail, it's definitely not the time to try touching that. Things are still far to different between the vehicles to do that at this point, but we can slowly work towards it (part of the reason we were pushing the recent copter failsafe stuff to use the FailsafeAction enum that battery failsafes do)

peterbarker commented:
On Thu, 24 Oct 2019, Tom Pittenger wrote:

> The concept is good. There's been talk about abstracting all the fail-safes into it's own library for all vehicles. Perhaps @peterbarker already has a
> branch for it?

I do - but there's not much in it.

https://github.com/ArduPilot/ardupilot/compare/master...peterbarker:wip/wind-failsafe?expand=1

rmackay9 commented:
We're planning the same kind of change in Rover as well.  we will split out all failsafes..

mday299 commented:
@WickedShell @peterbarker @magicrub @rmackay9 

I have updated this PR as follows:
- Not grabbing old indexes anymore
- Kept conversion code.
- Employed Failsafe_Action enum as @WickedShell suggested.
- Lowered maximum range limit in GCS failsafe timeout
- Remove bad whitespace change
- Removed type from printf

Let me know what you think.
